src/Containers/Microsoft.NET.Build.Containers/Tasks/CreateImageIndex.cs

```diff
@@ -0,0 +1,128 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System.Diagnostics;
+using Microsoft.Build.Framework;
+using Microsoft.Extensions.Logging;
+using Microsoft.NET.Build.Containers.Logging;
+using Microsoft.NET.Build.Containers.Resources;
+using ILogger = Microsoft.Extensions.Logging.ILogger;
+using Task = System.Threading.Tasks.Task;
+
+namespace Microsoft.NET.Build.Containers.Tasks;
+
+public sealed partial class CreateImageIndex : Microsoft.Build.Utilities.Task, ICancelableTask, IDisposable
+{
+    private readonly CancellationTokenSource _cancellationTokenSource = new();
+
+    public void Cancel() => _cancellationTokenSource.Cancel();
+
+    public void Dispose()
+    {
+        _cancellationTokenSource.Dispose();
+    }
+
+    public override bool Execute()
+    {
+        try
+        {
+            Task.Run(() => ExecuteAsync(_cancellationTokenSource.Token)).GetAwaiter().GetResult();
+        }
+        catch (TaskCanceledException ex)
+        {
+            Log.LogWarningFromException(ex);
+        }
+        catch (OperationCanceledException ex)
+        {
+            Log.LogWarningFromException(ex);
+        }
+        return !Log.HasLoggedErrors;
+    }
+
+    internal async Task<bool> ExecuteAsync(CancellationToken cancellationToken)
+    {
+        cancellationToken.ThrowIfCancellationRequested();
+
+        var images = ParseImages();
+        if (Log.HasLoggedErrors)
+        {
+            return !Log.HasLoggedErrors;
+        }
+
+        using MSBuildLoggerProvider loggerProvider = new(Log);
+        ILoggerFactory msbuildLoggerFactory = new LoggerFactory(new[] { loggerProvider });
+        ILogger logger = msbuildLoggerFactory.CreateLogger<CreateImageIndex>();
+
+        logger.LogInformation(Strings.BuildingImageIndex, GetRepositoryAndTagsString(), string.Join(", ", images.Select(i => i.ManifestDigest)));
+
+        try
+        {
+            (string imageIndex, string mediaType) = ImageIndexGenerator.GenerateImageIndex(images);
+
+            GeneratedImageIndex = imageIndex;
+
+            await PushToRemoteRegistry(GeneratedImageIndex, mediaType, logger, cancellationToken);
+        }
+        catch (ContainerHttpException e)
+        {
+            if (BuildEngine != null)
+            {
+                Log.LogErrorFromException(e, true);
+            }
+        }
+        catch (ArgumentException ex)
+        {
+            Log.LogErrorFromException(ex);
+        }
+
+        return !Log.HasLoggedErrors;
+    }
+
+    private ImageInfo[] ParseImages()
```

Comment:
Few files above, we have this information. Do we have to throw it away and then parse it again or does this have a different use case?
Also, if we want to enforce the media type to be consistent, do we have to store it with each image?
Why not store it just once and then throw is we encounter mismatch (which we already kind of do)?
It is perfectly possible that I'm missing something here.

Reply:
this is not what is happening here. 
when there is a case with multiple rids we call `_PublishMultiArchContainers` target. Inside of it, we call `_PublishSingleContainer` target per each rid, and gather outputs like manifest etc. in `GeneratedContainers` item (this is inside `_PublishSingleContainer` target). Then we call `CreateImageIndex` task with `ITaskItem[] GeneratedContainers` parameter. 
Here in `ParseImages` I parse them, and then validate the mediaType. There is no way to validate media types prior to this point.

Reply:
Thanks for the clarification